---
title: "Preprocessing GWAS-derived data"
author: "Fabricio Almeida-Silva"
date: "`r Sys.Date()`"
output: pdf_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo=TRUE,
                      warning=FALSE, 
                      message=FALSE,
                      tidy.opts=list(width.cutoff=60),
                      tidy=TRUE)
```

## Data description

- _snp_list.csv_: List of SNPs related to resistance to biotic stress along with reference to their original articles and the pathogen/pest to which the SNPs are related. This database was created by manually curating the scientific literature on genome-wide association studies that aimed at identifying loci affecting disease/pest resistance. The search strategy used on Google Scholar was: _intitle:"soybean" AND "GWAS" AND "resistance"_. Last update: March 2021.

## Data pre-processing and exploratory analyses

First, we must load the data and filter data to keep only SNPs related to *Ppachyrhizi*, *Mphaseolina*, *Fvirguliforme*, *Fgraminearum*, *Cgregata*.
```{r}
library(BioNERO)
library(here)
library(tidyverse)
species <- c("Ppachyrhizi", "Mphaseolina", "Fvirguliforme", 
             "Fgraminearum", "Cgregata")
snplist <- read.csv(here("data", "snp_list.csv"), header=TRUE, sep=";") %>%
  select(!c(1:3)) %>%
  distinct() %>%
  filter(Organism %in% species) %>%
  mutate(SNP = str_trim(SNP))
```

Now, let's count the number of markers by category.

```{r}
table(snplist$MarkerType)
```

To make the downstream analyses easier, let's separate metaQTLs from SNPs.

```{r}
metaQTL <- snplist[snplist$MarkerType == "metaQTL", ]
snplist <- snplist[snplist$MarkerType != "metaQTL", ] # Remove metaQTLs
head(snplist)
nrow(snplist)
length(unique(snplist$SNP))
```
There are **`r nrow(snplist)` SNPs** on this database. Of these, `r length(unique(snplist$SNP))` are unique, which means that there are 2 SNPs shared by different studies.

Let's see how many SNPs we have per genome version.

```{r}
table(snplist$GenomeVersion)
```

Now, counting the SNPs per organism.

```{r}
sort(table(snplist$Organism), decreasing=TRUE)
```

## Getting genome positions for the SNPs

The .vcf file containing SNP positions for *Wm82.a2.v1* was downloaded from Soybase.

Load the file with the SNP positions.

```{r}
positions_a2 <- read.csv(here("data", "snp_positions_a2.v1.txt"), 
                         header=T, sep = "\t")
names(positions_a2) <- c("chr", "pos", "id")
head(positions_a2)
```
Read in file with positions of SNPs starting with BARC*.

```{r}
barc_positions <- read.csv(here("data", "BARC_positions.txt"), 
                           header=TRUE, sep="\t")
head(barc_positions)
```

Combine `positions` and `barc_positions` in a single data frame. 

```{r}
pos1 <- positions_a2[, c(3,1,2, 2)]
names(pos1) <- c("SNP_ID", "Chr", "start", "end")
pos2 <- barc_positions[, c(1:4)]
names(pos2) <- names(pos1)
all_positions <- rbind(pos1, pos2)
head(all_positions)
```

Create new columns containing chromosome and position for the SNPs.

```{r}
snp_pos <- merge(snplist, all_positions, 
                 by.x="SNP", by.y="SNP_ID", all.x=TRUE)[, -c(5,6)]
head(snp_pos)
table(snp_pos$Chr)
```

Some SNPs do not have IDs. For these cases, SNP names are identified as `ChrX-pos`. Let's include genomic coordinates for these SNPs.

```{r}
sum(startsWith(snp_pos$SNP, "Chr"))
head(snp_pos[is.na(snp_pos$Chr), ])

# Add chromosome info
snp_pos$Chr[is.na(snp_pos$Chr)] <- sapply(strsplit(snp_pos$SNP[is.na(snp_pos$Chr)], "-"), `[`, 1)

# Add start and end positions
snp_pos$start[is.na(snp_pos$start)] <- sapply(
  strsplit(snp_pos$SNP[is.na(snp_pos$start)], "-"),
  `[`, 2
  ) 
snp_pos$end[is.na(snp_pos$end)] <- sapply(
  strsplit(snp_pos$SNP[is.na(snp_pos$end)], "-"), 
  tail, 1
  ) 

# Check if they have all been replaced
nrow(snp_pos[rowSums(is.na(snp_pos)) > 0, ])
```

To conclude, let's standardize chromosome names based on the `ChrX` nomenclature and save the SNP positions as GRanges and GRangesList objects.

```{r save_snp_granges}
library(GenomicRanges)
snp_ranges <- snp_pos %>%
  filter(!startsWith(Chr, "scaffold")) %>%
  mutate(Chr = str_replace(Chr, "Gm", "Chr")) %>%
  mutate(Chr = str_replace_all(Chr, 
                           c("Chr2$" = "Chr02", 
                             "Chr4" = "Chr04", 
                             "Chr5" = "Chr05", 
                             "Chr6" = "Chr06", 
                             "Chr7" = "Chr07"))
         ) %>%
  select(Chr, start, end, Organism, Trait, SNP, DOI)

# GRanges
snp_granges <- makeGRangesFromDataFrame(snp_granges)
snp_granges$Organism <- snp_ranges$Organism
snp_granges$Trait <- snp_ranges$Trait
snp_granges$SNP <- snp_ranges$SNP
snp_granges$DOI <- snp_ranges$DOI

# GRangesList
snp_grangeslist <- split(snp_granges, snp_granges$Organism)

# Saving data
save(snp_granges, snp_grangeslist, 
     file = here("products", "result_files", "snp_granges.rda"),
     compress = "xz")
```

To wrap it up, making a table of studies included in this study. This will be Table 1 in the manuscript.

```{r table1}
table1 <- as.data.frame(snp_granges)
table1 %>%
  count(DOI, Organism) %>%
  arrange(DOI, Organism)
```

## Session information

This document was created under the following conditions:

```{r}
sessionInfo()
```